src/main/java/duke/main/Parser.java

```diff
@@ -0,0 +1,118 @@
+package duke.main;
+
+import duke.exception.EmptyTaskException;
+import duke.exception.InvalidCommandException;
+import duke.exception.InvalidDateException;
+import duke.exception.InvalidIndexException;
+import duke.exception.MissingDateException;
+import duke.task.Task;
+
+public class Parser {
+    private final TaskList taskList;
+    private StringBuilder reply = new StringBuilder();
+
+    Parser(TaskList taskList) {
+        this.taskList = taskList;
+    }
+
+    /**
+     * Interprets user input and executes commands.
+     * The method supports commands for adding, deleting, viewing and completing tasks.
+     * If invalid input is given, users are informed and told to try again.
+     *
+     * @param input User command in console.
+     * @return Output message upon execution of command.
+     */
+    String parse(String input) {
+        try {
+            if (input.equals("list")) {
+                list();
+            } else if (input.startsWith("done")) {
+                done(input);
+            } else if (input.startsWith("delete")) {
+                delete(input);
+            } else if (input.startsWith("find")) {
+                find(input);
+            } else {
+                add(input);
+            }
+        } catch (InvalidCommandException e) {
+            reply.append(e.toString());
+        }
+        StringBuilder output = reply;
+        reply = new StringBuilder();
+        return output.toString();
+    }
+
+    private void list() {
+        reply.append("Here are the tasks in your list:")
+                .append(System.lineSeparator())
+                .append(taskList);
+    }
+
+    private void done(String input) throws InvalidCommandException {
+        if (input.length() != 6) {
+            throw new InvalidCommandException();
+        } else {
+            try {
+                int index = Integer.parseInt(input.substring(5)) - 1;
+                Task completed = taskList.complete(index);
+                reply.append("Well done! The following task is complete:")
+                        .append(System.lineSeparator())
+                        .append(completed);
+            } catch (InvalidIndexException e) {
+                reply.append(e.toString());
+            }
+        }
+    }
+
+    private void delete(String input) throws InvalidCommandException {
+        if (input.length() != 8) {
+            throw new InvalidCommandException();
+        } else {
+            try {
+                int index = Integer.parseInt(input.substring(7)) - 1;
+                Task deleted = taskList.delete(index);
+                reply.append("Noted, I have removed the below task:")
```

Comment:
I would recommend concatenating the string or using your StringBuilder to construct your input as a whole instead of appending so many times